Practical Example/RWTakeHome/RWTakeHome/ViewControllers/TutorialViewController.swift

```diff
@@ -0,0 +1,123 @@
+  @IBAction func segmentedControlValueChanged(_ sender: UISegmentedControl) {
+    
+    switch sender.selectedSegmentIndex {
+    case 0: // Article
+
+      self.tutorials = self.articles.sorted(by: <)
+
+      self.reloadDataOnMainThread()
+
+    case 1: // Video
+
+      self.tutorials = self.videos.sorted(by: <)
+
+      self.reloadDataOnMainThread()
+
+    default: // All
+
+      self.tutorials = self.videos.sorted(by: <) + self.articles.sorted(by: <)
+      
+      self.reloadDataOnMainThread()
+    }
+  }
+  //MARK:- View Life Cycle
+
+  override func viewDidLoad() {
+    super.viewDidLoad()
+
+    getArticles()
+    getVideos()
+
+  }
+}
+//MARK:- UICollection Datasource & Delegate
+extension TutorialViewController: UICollectionViewDelegate, UICollectionViewDataSource {
+  func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+    return tutorials.count
+  }
+
+  func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+    let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "TutorialCell", for: indexPath) as! TutorialCell
+
+    let article = self.tutorials[indexPath.row]
+
+    cell.descriptionLabel.text = article.attributes.description
+
+    cell.releaseDate.text = "Release Date: \(article.attributes.releaseDate.toString())"
+
+    cell.nameLabel.text = article.attributes.name
+
+    cell.typeLabel.text = article.attributes.contentType == "collection" ? "Video" : article.attributes.contentType.capitalized
+
+    loadImage(forCell: cell, forItemAt: indexPath)
+
+    return cell
+  }
+  private func loadImage(forCell cell: TutorialCell, forItemAt indexPath: IndexPath) {
+    let tutorial = tutorials[indexPath.item]
+
+    // Check for image in cache
+    if let cachedImageData = cache.value(for: tutorial.attributes.name),
+       let image = UIImage(data: cachedImageData) {
+      cell.artworkImageView.image = image
+      return
+    }
+
+    // Start an operation to fetch image data
+    let fetchOp = FetchPhotoOperation(tutorial: tutorial)
+    
+    let cacheOp = BlockOperation {
+      if let data = fetchOp.imageData {
+        self.cache.cache(value: data, for: tutorial.attributes.name)
+      }
+    }
+    let completionOp = BlockOperation {
+      defer { self.operations.removeValue(forKey: tutorial.attributes.name) }
+
+      if let currentIndexPath = self.collectionView?.indexPath(for: cell),
+         currentIndexPath != indexPath {
+        print("Got image for now-reused cell")
+        return
+      }
+
+      if let data = fetchOp.imageData {
+        cell.artworkImageView.image = UIImage(data: data)
+      }
+    }
+
+    cacheOp.addDependency(fetchOp)
+    completionOp.addDependency(fetchOp)
+
+    photoFetchQueue.addOperation(fetchOp)
+    photoFetchQueue.addOperation(cacheOp)
+    OperationQueue.main.addOperation(completionOp)
+
+    operations[tutorial.attributes.name] = fetchOp
+  }
```

Comment:
It's cool you made your own implementation of image background fetching and caching. Really amazing work! One thing I'd suggest is that you cancel fetches as well. If the user quickly scrolls through the collection view, you might start fetching 20 images at once. You'll waste a lot of bandwidth but also slow down fetching with each new scroll. You can use collection view delegate and data source methods to get notified when a cell stops being visible. You'll also need a way to store currently running operations (perhaps a dictionary of [Index Path:: Operation]) to cancel the appropriate one.